lang/py/src/avro/io.py

```diff
@@ -280,6 +298,83 @@ def read_utf8(self):
     """
     return unicode(self.read_bytes(), "utf-8")
 
+  def read_date_from_int(self):
+    """
+    Decode python date object from int
+    int stores the number of days from
+    the unix epoch, 1 January 1970 (ISO calendar).
+    """
+    days_since_epoc = self.read_int()
+    if days_since_epoc < 0:
+      raise schema.AvroException("Cannot decode date before 1 January 1970 (ISO calendar)")
+    return datetime.date(1970, 1, 1) + datetime.timedelta(days_since_epoc)
+
+  def _build_time_object(self, value, unit_multiplier):
+    value = value * (1000 // unit_multiplier)
+    value, microseconds =  value // 1000000, value % 1000000
+    value, seconds = value // 60, value % 60
+    value, minutes = value // 60, value % 60
+    hours = value
+
+    return datetime.time(
+      hour=hours,
+      minute=minutes,
+      second=seconds,
+      microsecond=microseconds
+    )
+
+  def _validate_time_datum(self, value, time_type):
+    if (time_type in [constants.TIME_MILLIS] and
+        not (UNSIGNED_INT_MIN_VALUE <= value <= UNSIGNED_INT_MAX_VALUE)):
+      raise schema.AvroException("Cannot decode %s logicaltype schema with value outside (%d, %d)." % (
+        time_type,
+        UNSIGNED_INT_MIN_VALUE,
+        UNSIGNED_INT_MAX_VALUE)
+      )
+    elif (time_type in [constants.TIME_MICROS, constants.TIMESTAMP_MILLIS, constants.TIMESTAMP_MICROS] and
```

Comment:
I think the timestamp can be negative which represents the timeline before 1/1/1970.